src/libcephd/Objecter.cc

```diff
@@ -0,0 +1,487 @@
+  boost::uuids::uuid vol;
+  epoch_t epoch = 0;
+  mempcpy(&vol, volume, sizeof(vol));
+  int64_t pool_id(0);
+  object_locator_t oloc(pool_id);
+  pg_t pgid;
+
+  // when asynchronous, flags must specify one or more of UNSTABLE or STABLE
+  if ((flags & WRITE_CB_FLAGS) == 0)
+    return -EINVAL;
+
+  if (!wait_for_active(&epoch))
+    return -ENODEV;
+
+  bufferlist bl;
+  bl.append(ceph::buffer::create_static(length, data));
+
+  // set up osd write op
+  MOSDOp *m =
+    new MOSDOp(client, tid, oid, oloc, pgid, epoch, flags,0);
+  m->set_snapid(CEPH_NOSNAP);
+  vector<OSDOp> ops;
+  ops.resize(1);
+  ops[0].op.op = CEPH_OSD_OP_WRITE;
+  ops[0].op.extent.offset = offset;
+  ops[0].op.extent.length = length;
+  ops[0].op.extent.truncate_size = 0;
+  ops[0].op.extent.truncate_seq = 0;
+  ops[0].indata = bl;
+  ops[0].op.flags = flags;
+  m->ops=ops;
+  
+  if (flags & LIBOSD_WRITE_CB_UNSTABLE)
+    m->set_want_ack(true);
+  if (flags & LIBOSD_WRITE_CB_STABLE)
+    m->set_want_ondisk(true);
+
+  // create reply callback
+  OnWriteReply *onreply = new OnWriteReply(cb, flags, user);
+
+  // send request over direct messenger
+  dispatcher->send_request(m, onreply);
+  return 0;
+}
+
+  int Objecter::truncate_sync(const char *object, const int64_t pool_id,
+			      OSDMapRef &omap,uint64_t offset, int flags)
+{
+  const int client = 0;
+  const long tid = 0;
+  object_t oid(object);
+  epoch_t epoch = 0;
+  SyncCompletion completion;
+  object_locator_t oloc(pool_id);
+  pg_t pgid;
+
+  omap->object_locator_to_pg(oid, oloc, pgid);
+
+  // when synchronous, flags must specify exactly one of UNSTABLE or STABLE
+  if ((flags & WRITE_CB_FLAGS) == 0 ||
+      (flags & WRITE_CB_FLAGS) == WRITE_CB_FLAGS)
+    return -EINVAL;
+
+  if (!wait_for_active(&epoch))
+    return -ENODEV;
+
+  // set up osd truncate op
+  MOSDOp *m =
+    new MOSDOp(client, tid, oid, oloc, pgid, epoch, flags,0);
+  vector<OSDOp> ops;
+  ops.resize(1);
+  ops[0].op.op = CEPH_OSD_OP_TRUNCATE;
+  ops[0].op.extent.offset = offset;
+  ops[0].op.extent.truncate_size = 0;
+  ops[0].op.extent.truncate_seq = 0;
+  ops[0].op.flags = flags;
+  m->ops=ops;
+
+  if (flags & LIBOSD_WRITE_CB_UNSTABLE)
+    m->set_want_ack(true);
+  if (flags & LIBOSD_WRITE_CB_STABLE)
+    m->set_want_ondisk(true);
+
+  // create reply callback on the stack
+  OnWriteReply onreply(SyncCompletion::callback, flags, &completion);
+
+  // send request over direct messenger
+  dispatcher->send_request(m, &onreply);
+
+  return completion.wait();
+
+}
+
+  int Objecter::truncate(const char *object, const int64_t pool_id,
+			 OSDMapRef& omap, uint64_t offset, int flags,
+		       libosd_io_completion_fn cb, void *user)
+{
+  if (cb == nullptr)
+    return truncate_sync(object, pool_id, omap, offset, flags);
+
+  const int client = 0;
+  const long tid = 0;
+  object_t oid(object);
+  epoch_t epoch = 0;
+  object_locator_t oloc(pool_id);
+  pg_t pgid;
+
+  omap->object_locator_to_pg(oid, oloc, pgid);
```

Comment:
this call should be in the read and write paths also (they're both passing an empty pgid now)

consider passing this pgid into `truncate_sync()`, so we don't have to duplicate the call there